[View in Colaboratory](https://colab.research.google.com/github/koyopro/kaggle/blob/master/recruit/recruit.ipynb)

# Recruit Restaurant Visitor Forecasting

https://www.kaggle.com/c/recruit-restaurant-visitor-forecasting  
2018/05/17

- 予約数から客数を予測する問題
- お店毎に2017/4/23~2017/5/31の数を予測する？
  - ゴールデンウィークを含むことに注意
- 日付データは2016/01/01~2017/05/31
- お店は全150店舗
- 学習データ(実際の来店客数)はair_visit_data.csv
  - 2016/01~2017/04/22までのデータ
- 学習データとテストデータはairのIDなので、先にhpgは無視してairだけでやってみるのもアリか。

In [0]:
import pandas as pd
import os
import requests
import xgboost as xgb
import numpy as np
from sklearn import preprocessing
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline 
from google.colab import files

In [2]:
open('storage.py', 'wb').write(requests.get('https://www.dropbox.com/s/xaohy7hkltic6kv/storage.py?dl=1', allow_redirects=True).content)
import storage
storage.sync_from_storage('common')

# !pip install kaggle

gsutil rsync -r gs://staging.kaggle-ky.appspot.com/common .



In [3]:
ls

air_reserve.csv     date_info.csv       sample_submission.csv
air_store_info.csv  hpg_reserve.csv     storage.py
air_visit_data.csv  hpg_store_info.csv  store_id_relation.csv
datalab/            __pycache__/


In [4]:
storage.sync_from_storage('recruit')

gsutil rsync -r gs://staging.kaggle-ky.appspot.com/recruit .



In [5]:
ls

air_reserve.csv     date_info.csv       sample_submission.csv
air_store_info.csv  hpg_reserve.csv     storage.py
air_visit_data.csv  hpg_store_info.csv  store_id_relation.csv
datalab/            __pycache__/


In [0]:
submission = pd.read_csv('sample_submission.csv')

In [7]:
submission.head()

id  visitors
0  air_00a91d42b08b08d9_2017-04-23         0
1  air_00a91d42b08b08d9_2017-04-24         0
2  air_00a91d42b08b08d9_2017-04-25         0
3  air_00a91d42b08b08d9_2017-04-26         0
4  air_00a91d42b08b08d9_2017-04-27         0

In [8]:
submission.tail()

id  visitors
32014  air_fff68b929994bfbd_2017-05-27         0
32015  air_fff68b929994bfbd_2017-05-28         0
32016  air_fff68b929994bfbd_2017-05-29         0
32017  air_fff68b929994bfbd_2017-05-30         0
32018  air_fff68b929994bfbd_2017-05-31         0

In [0]:
date_info = pd.read_csv('date_info.csv')

In [10]:
date_info.head()

calendar_date day_of_week  holiday_flg
0    2016-01-01      Friday            1
1    2016-01-02    Saturday            1
2    2016-01-03      Sunday            1
3    2016-01-04      Monday            0
4    2016-01-05     Tuesday            0

In [11]:
date_info.tail()

calendar_date day_of_week  holiday_flg
512    2017-05-27    Saturday            0
513    2017-05-28      Sunday            0
514    2017-05-29      Monday            0
515    2017-05-30     Tuesday            0
516    2017-05-31   Wednesday            0

In [12]:
air_reserve = pd.read_csv('air_reserve.csv')
air_reserve.head()

air_store_id       visit_datetime     reserve_datetime  \
0  air_877f79706adbfb06  2016-01-01 19:00:00  2016-01-01 16:00:00   
1  air_db4b38ebe7a7ceff  2016-01-01 19:00:00  2016-01-01 19:00:00   
2  air_db4b38ebe7a7ceff  2016-01-01 19:00:00  2016-01-01 19:00:00   
3  air_877f79706adbfb06  2016-01-01 20:00:00  2016-01-01 16:00:00   
4  air_db80363d35f10926  2016-01-01 20:00:00  2016-01-01 01:00:00   

   reserve_visitors  
0                 1  
1                 3  
2                 6  
3                 2  
4                 5

In [13]:
air_reserve.tail()

air_store_id       visit_datetime     reserve_datetime  \
92373  air_db4b38ebe7a7ceff  2017-05-29 18:00:00  2017-04-17 19:00:00   
92374  air_1033310359ceeac1  2017-05-30 23:00:00  2017-04-19 17:00:00   
92375  air_877f79706adbfb06  2017-05-31 20:00:00  2017-04-12 23:00:00   
92376  air_900d755ebd2f7bbd  2017-05-31 20:00:00  2017-04-19 16:00:00   
92377  air_3cad29d1a23209d2  2017-05-31 21:00:00  2017-04-22 10:00:00   

       reserve_visitors  
92373                 6  
92374                 8  
92375                 3  
92376                10  
92377                 3

In [0]:
hpg_reserve = pd.read_csv('hpg_reserve.csv')

In [15]:
hpg_reserve.head()

hpg_store_id       visit_datetime     reserve_datetime  \
0  hpg_c63f6f42e088e50f  2016-01-01 11:00:00  2016-01-01 09:00:00   
1  hpg_dac72789163a3f47  2016-01-01 13:00:00  2016-01-01 06:00:00   
2  hpg_c8e24dcf51ca1eb5  2016-01-01 16:00:00  2016-01-01 14:00:00   
3  hpg_24bb207e5fd49d4a  2016-01-01 17:00:00  2016-01-01 11:00:00   
4  hpg_25291c542ebb3bc2  2016-01-01 17:00:00  2016-01-01 03:00:00   

   reserve_visitors  
0                 1  
1                 3  
2                 2  
3                 5  
4                13

In [16]:
hpg_reserve.tail()

hpg_store_id       visit_datetime     reserve_datetime  \
2000315  hpg_2b293477ea4642fb  2017-05-31 20:00:00  2017-04-19 07:00:00   
2000316  hpg_f922aeca48ca5a29  2017-05-31 20:00:00  2017-03-19 00:00:00   
2000317  hpg_e9151de687b93aa3  2017-05-31 21:00:00  2017-04-14 20:00:00   
2000318  hpg_fae6c96189b4a122  2017-05-31 21:00:00  2017-04-19 17:00:00   
2000319  hpg_0b70de808b55ad1e  2017-05-31 23:00:00  2017-04-21 13:00:00   

         reserve_visitors  
2000315                 2  
2000316                 3  
2000317                49  
2000318                 9  
2000319                 6

In [0]:
store_id_relation = pd.read_csv('store_id_relation.csv')

In [18]:
store_id_relation.head()

air_store_id          hpg_store_id
0  air_63b13c56b7201bd9  hpg_4bc649e72e2a239a
1  air_a24bf50c3e90d583  hpg_c34b496d0305a809
2  air_c7f78b4f3cba33ff  hpg_cd8ae0d9bbd58ff9
3  air_947eb2cae4f3e8f2  hpg_de24ea49dc25d6b8
4  air_965b2e0cf4119003  hpg_653238a84804d8e7

In [19]:
store_id_relation.shape

(150, 2)

In [0]:
air_visit_data = pd.read_csv('air_visit_data.csv')

In [21]:
air_visit_data.head()

air_store_id  visit_date  visitors
0  air_ba937bf13d40fb24  2016-01-13        25
1  air_ba937bf13d40fb24  2016-01-14        32
2  air_ba937bf13d40fb24  2016-01-15        29
3  air_ba937bf13d40fb24  2016-01-16        22
4  air_ba937bf13d40fb24  2016-01-18         6

In [22]:
air_visit_data.tail()

air_store_id  visit_date  visitors
252103  air_24e8414b9b07decb  2017-04-18         6
252104  air_24e8414b9b07decb  2017-04-19         6
252105  air_24e8414b9b07decb  2017-04-20         7
252106  air_24e8414b9b07decb  2017-04-21         8
252107  air_24e8414b9b07decb  2017-04-22         5

In [23]:
dates = pd.DataFrame(np.unique(air_visit_data['visit_date']))
print(dates.head())
print(dates.tail())

            0
0  2016-01-01
1  2016-01-02
2  2016-01-03
3  2016-01-04
4  2016-01-05
              0
473  2017-04-18
474  2017-04-19
475  2017-04-20
476  2017-04-21
477  2017-04-22


In [0]:
air_store_info = pd.read_csv('air_store_info.csv')

In [25]:
air_store_info.head()

air_store_id  air_genre_name                 air_area_name  \
0  air_0f0cdeee6c9bf3d7  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
1  air_7cc17a324ae5c7dc  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
2  air_fee8dcf4d619598e  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
3  air_a17f0778617c76e2  Italian/French  Hyōgo-ken Kōbe-shi Kumoidōri   
4  air_83db5aff8f50478e  Italian/French  Tōkyō-to Minato-ku Shibakōen   

    latitude   longitude  
0  34.695124  135.197852  
1  34.695124  135.197852  
2  34.695124  135.197852  
3  34.695124  135.197852  
4  35.658068  139.751599

In [26]:
air_store_info.tail()

air_store_id         air_genre_name  \
824  air_9bf595ef095572fb  International cuisine   
825  air_764f71040a413d4d                  Asian   
826  air_10bbe8acd943d8f6                  Asian   
827  air_7514d90009613cd6          Karaoke/Party   
828  air_c8fe396d6c46275d          Karaoke/Party   

                             air_area_name   latitude   longitude  
824            Tōkyō-to Shibuya-ku Shibuya  35.661777  139.704051  
825            Tōkyō-to Shibuya-ku Shibuya  35.661777  139.704051  
826            Tōkyō-to Shibuya-ku Shibuya  35.661777  139.704051  
827  Hokkaidō Sapporo-shi Minami 3 Jōnishi  43.055460  141.340956  
828  Hokkaidō Sapporo-shi Minami 3 Jōnishi  43.055460  141.340956

In [27]:
air_store_info.shape

(829, 5)

In [0]:
hpg_store_info = pd.read_csv('hpg_store_info.csv')

In [29]:
hpg_store_info.head()

hpg_store_id  hpg_genre_name                  hpg_area_name  \
0  hpg_6622b62385aec8bf  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
1  hpg_e9e068dd49c5fa00  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
2  hpg_2976f7acb4b3a3bc  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
3  hpg_e51a522e098f024c  Japanese style  Tōkyō-to Setagaya-ku Taishidō   
4  hpg_e3d0e1519894f275  Japanese style  Tōkyō-to Setagaya-ku Taishidō   

    latitude   longitude  
0  35.643675  139.668221  
1  35.643675  139.668221  
2  35.643675  139.668221  
3  35.643675  139.668221  
4  35.643675  139.668221

In [30]:
hpg_store_info.tail()

hpg_store_id               hpg_genre_name  \
4685  hpg_c6ed353a48c7c93e                 Sichuan food   
4686  hpg_a58d2423360b2fb4                 Sichuan food   
4687  hpg_d2021704076cdd70                Shanghai food   
4688  hpg_602472570d65e219  Spain/Mediterranean cuisine   
4689  hpg_11801306ce8bfb4f                    Udon/Soba   

                    hpg_area_name   latitude   longitude  
4685     Tōkyō-to Chiyoda-ku None  35.695780  139.768453  
4686  Tōkyō-to Minato-ku Roppongi  35.666137  139.733978  
4687     Tōkyō-to Chiyoda-ku None  35.695780  139.768453  
4688     Tōkyō-to Toshima-ku None  35.730068  139.711742  
4689  Osaka Prefecture Osaka None  34.701519  135.498859

hpg_store_id多くない…？
よく見たらair_store_idも828あるわ。
そういうことか。air_storeの828が正だけど、hpg_storeにも入ってるのはそのうち150店だけで、
予測と関係ない店も4500件くらい入ってるわけね…。

とりあえずどんな特徴でやるか。  
- air_store_id
- 休日の何日前か(休日なら0)
- 前日のvisitor数
- 前回の休日のvisitor数
- 前回の平日のvisitor数

よく見たらholiday_flgは祝日だけ1なんやな

In [31]:
date_info = pd.read_csv('date_info.csv')
date_info.head()

calendar_date day_of_week  holiday_flg
0    2016-01-01      Friday            1
1    2016-01-02    Saturday            1
2    2016-01-03      Sunday            1
3    2016-01-04      Monday            0
4    2016-01-05     Tuesday            0

In [0]:
_df = pd.DataFrame({
    'calendar_date': ['2017-06-01', '2017-06-02', '2017-06-03', '2017-06-04', '2017-06-05', '2017-06-06', '2017-06-07'],
    'day_of_week': ['Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday', 'Monday'],
    'holiday_flg': [0, 0, 0, 0, 1, 1, 0],
})
date_info = pd.concat([date_info, _df]).reset_index(drop=True)

In [0]:
saturday = date_info['day_of_week'] == 'Saturday'
sunday = date_info['day_of_week'] == 'Sunday'
date_info.loc[saturday | sunday, 'holiday_flg'] = 1

In [0]:
for i in range(7):
  key = f'{i}_days_to_holiday'
  date_info[key] = date_info['holiday_flg'][i:].reset_index(drop=True).astype('bool')
  date_info[key] = date_info[key].astype('bool')

In [35]:
print(date_info.shape)
date_info.dtypes

(524, 10)


calendar_date        object
day_of_week          object
holiday_flg           int64
0_days_to_holiday      bool
1_days_to_holiday      bool
2_days_to_holiday      bool
3_days_to_holiday      bool
4_days_to_holiday      bool
5_days_to_holiday      bool
6_days_to_holiday      bool
dtype: object

In [36]:
date_info.head(10)

calendar_date day_of_week  holiday_flg  0_days_to_holiday  \
0    2016-01-01      Friday            1               True   
1    2016-01-02    Saturday            1               True   
2    2016-01-03      Sunday            1               True   
3    2016-01-04      Monday            0              False   
4    2016-01-05     Tuesday            0              False   
5    2016-01-06   Wednesday            0              False   
6    2016-01-07    Thursday            0              False   
7    2016-01-08      Friday            0              False   
8    2016-01-09    Saturday            1               True   
9    2016-01-10      Sunday            1               True   

   1_days_to_holiday  2_days_to_holiday  3_days_to_holiday  4_days_to_holiday  \
0               True               True              False              False   
1               True              False              False              False   
2              False              False              False              False   
3              False              False              False              False   
4              False              False              False               True   
5              False              False               True               True   
6              False               True               True               True   
7               True               True               True              False   
8               True               True              False              False   
9               True              False              False              False   

   5_days_to_holiday  6_days_to_holiday  
0              False              False  
1              False              False  
2              False               True  
3               True               True  
4               True               True  
5               True              False  
6              False              False  
7              False              False  
8              False              False  
9              False               True

In [37]:
date_info.tail(10)

calendar_date day_of_week  holiday_flg  0_days_to_holiday  \
514    2017-05-29      Monday            0              False   
515    2017-05-30     Tuesday            0              False   
516    2017-05-31   Wednesday            0              False   
517    2017-06-01     Tuesday            0              False   
518    2017-06-02   Wednesday            0              False   
519    2017-06-03    Thursday            0              False   
520    2017-06-04      Friday            0              False   
521    2017-06-05    Saturday            1               True   
522    2017-06-06      Sunday            1               True   
523    2017-06-07      Monday            0              False   

     1_days_to_holiday  2_days_to_holiday  3_days_to_holiday  \
514              False              False              False   
515              False              False              False   
516              False              False              False   
517              False              False              False   
518              False              False               True   
519              False               True               True   
520               True               True              False   
521               True              False               True   
522              False               True               True   
523               True               True               True   

     4_days_to_holiday  5_days_to_holiday  6_days_to_holiday  
514              False              False              False  
515              False              False               True  
516              False               True               True  
517               True               True              False  
518               True              False               True  
519              False               True               True  
520               True               True               True  
521               True               True               True  
522               True               True               True  
523               True               True               True

date_infoの準備OK

In [0]:
train = air_visit_data.copy()

In [39]:
print(train.shape)
train.head()

(252108, 3)


air_store_id  visit_date  visitors
0  air_ba937bf13d40fb24  2016-01-13        25
1  air_ba937bf13d40fb24  2016-01-14        32
2  air_ba937bf13d40fb24  2016-01-15        29
3  air_ba937bf13d40fb24  2016-01-16        22
4  air_ba937bf13d40fb24  2016-01-18         6

In [40]:
test = pd.read_csv('sample_submission.csv')
test.head()

id  visitors
0  air_00a91d42b08b08d9_2017-04-23         0
1  air_00a91d42b08b08d9_2017-04-24         0
2  air_00a91d42b08b08d9_2017-04-25         0
3  air_00a91d42b08b08d9_2017-04-26         0
4  air_00a91d42b08b08d9_2017-04-27         0

In [41]:
test = submission['id'].str.split('_', expand=True)
test.columns = ['air', 'store_id', 'visit_date']
test['air_store_id'] = test['air'] + '_' + test['store_id']
test = test.drop(columns=['air', 'store_id'])
print(test.shape)
test.head()

(32019, 2)


visit_date          air_store_id
0  2017-04-23  air_00a91d42b08b08d9
1  2017-04-24  air_00a91d42b08b08d9
2  2017-04-25  air_00a91d42b08b08d9
3  2017-04-26  air_00a91d42b08b08d9
4  2017-04-27  air_00a91d42b08b08d9

In [42]:
all_data = pd.concat([train.drop(columns=['visitors']), test])
print(all_data.shape)
all_data.head()

(284127, 2)


air_store_id  visit_date
0  air_ba937bf13d40fb24  2016-01-13
1  air_ba937bf13d40fb24  2016-01-14
2  air_ba937bf13d40fb24  2016-01-15
3  air_ba937bf13d40fb24  2016-01-16
4  air_ba937bf13d40fb24  2016-01-18

In [43]:
all_data = pd.merge(all_data, date_info, how='left', left_on='visit_date', right_on='calendar_date')
all_data = all_data.drop(columns=['visit_date', 'calendar_date'])
print(all_data.shape)
all_data.head()

(284127, 10)


air_store_id day_of_week  holiday_flg  0_days_to_holiday  \
0  air_ba937bf13d40fb24   Wednesday            0              False   
1  air_ba937bf13d40fb24    Thursday            0              False   
2  air_ba937bf13d40fb24      Friday            0              False   
3  air_ba937bf13d40fb24    Saturday            1               True   
4  air_ba937bf13d40fb24      Monday            0              False   

   1_days_to_holiday  2_days_to_holiday  3_days_to_holiday  4_days_to_holiday  \
0              False              False               True               True   
1              False               True               True              False   
2               True               True              False              False   
3               True              False              False              False   
4              False              False              False              False   

   5_days_to_holiday  6_days_to_holiday  
0              False              False  
1              False              False  
2              False              False  
3              False              False  
4               True               True

In [44]:
all_data.dtypes

air_store_id         object
day_of_week          object
holiday_flg           int64
0_days_to_holiday      bool
1_days_to_holiday      bool
2_days_to_holiday      bool
3_days_to_holiday      bool
4_days_to_holiday      bool
5_days_to_holiday      bool
6_days_to_holiday      bool
dtype: object

学習用データの完成(後はget_dummiesをやる)

In [45]:
X = pd.get_dummies(all_data)
X.head()

holiday_flg  0_days_to_holiday  1_days_to_holiday  2_days_to_holiday  \
0            0              False              False              False   
1            0              False              False               True   
2            0              False               True               True   
3            1               True               True              False   
4            0              False              False              False   

   3_days_to_holiday  4_days_to_holiday  5_days_to_holiday  6_days_to_holiday  \
0               True               True              False              False   
1               True              False              False              False   
2              False              False              False              False   
3              False              False              False              False   
4              False              False               True               True   

   air_store_id_air_00a91d42b08b08d9  air_store_id_air_0164b9927d20bcc3  \
0                                  0                                  0   
1                                  0                                  0   
2                                  0                                  0   
3                                  0                                  0   
4                                  0                                  0   

           ...            air_store_id_air_fef9ccb3ba0da2f7  \
0          ...                                            0   
1          ...                                            0   
2          ...                                            0   
3          ...                                            0   
4          ...                                            0   

   air_store_id_air_ffcc2d5087e1b476  air_store_id_air_fff68b929994bfbd  \
0                                  0                                  0   
1                                  0                                  0   
2                                  0                                  0   
3                                  0                                  0   
4                                  0                                  0   

   day_of_week_Friday  day_of_week_Monday  day_of_week_Saturday  \
0                   0                   0                     0   
1                   0                   0                     0   
2                   1                   0                     0   
3                   0                   0                     1   
4                   0                   1                     0   

   day_of_week_Sunday  day_of_week_Thursday  day_of_week_Tuesday  \
0                   0                     0                    0   
1                   0                     1                    0   
2                   0                     0                    0   
3                   0                     0                    0   
4                   0                     0                    0   

   day_of_week_Wednesday  
0                      1  
1                      0  
2                      0  
3                      0  
4                      0  

[5 rows x 844 columns]

In [46]:
print(X.dtypes[:10], '\n\n', X.dtypes[-10:])

holiday_flg                          int64
0_days_to_holiday                     bool
1_days_to_holiday                     bool
2_days_to_holiday                     bool
3_days_to_holiday                     bool
4_days_to_holiday                     bool
5_days_to_holiday                     bool
6_days_to_holiday                     bool
air_store_id_air_00a91d42b08b08d9    uint8
air_store_id_air_0164b9927d20bcc3    uint8
dtype: object 

 air_store_id_air_fef9ccb3ba0da2f7    uint8
air_store_id_air_ffcc2d5087e1b476    uint8
air_store_id_air_fff68b929994bfbd    uint8
day_of_week_Friday                   uint8
day_of_week_Monday                   uint8
day_of_week_Saturday                 uint8
day_of_week_Sunday                   uint8
day_of_week_Thursday                 uint8
day_of_week_Tuesday                  uint8
day_of_week_Wednesday                uint8
dtype: object


In [47]:
X_train = X[:train.shape[0]]
X_test = X[train.shape[0]:]
print(X_train.shape, X_test.shape)

(252108, 844) (32019, 844)


In [48]:
y_train = train['visitors']
print(y_train.shape)

(252108,)


In [49]:
model = xgb.XGBRegressor()
model.fit(X_train, y_train)

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
       n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=True, subsample=1)

In [0]:
predict = model.predict(X_test)

In [52]:
predict.shape

(32019,)

In [53]:
output = submission.copy()
output.head()

id  visitors
0  air_00a91d42b08b08d9_2017-04-23         0
1  air_00a91d42b08b08d9_2017-04-24         0
2  air_00a91d42b08b08d9_2017-04-25         0
3  air_00a91d42b08b08d9_2017-04-26         0
4  air_00a91d42b08b08d9_2017-04-27         0

In [0]:
output.visitors = predict

In [55]:
output.head()

id   visitors
0  air_00a91d42b08b08d9_2017-04-23  22.199228
1  air_00a91d42b08b08d9_2017-04-24  17.394461
2  air_00a91d42b08b08d9_2017-04-25  17.774137
3  air_00a91d42b08b08d9_2017-04-26  17.944056
4  air_00a91d42b08b08d9_2017-04-27  17.944056

In [0]:
output.to_csv('output.csv', index=False)

In [60]:
ls

air_reserve.csv     datalab/         hpg_store_info.csv  sample_submission.csv
air_store_info.csv  date_info.csv    output.csv          storage.py
air_visit_data.csv  hpg_reserve.csv  __pycache__/        store_id_relation.csv


In [0]:
files.download('output.csv')